File: Problem_2.js
// Matchsticks to Square (https://leetcode.com/problems/matchsticks-to-square/)

// TC: O(4^N)
// SC: O(N)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

// TC: 4^N
// SC: N

let sums, possibleSumSide;
/**
 * @param {number[]} matchsticks
 * @return {boolean}
 */
var makesquare = function (matchsticks) {
    if (matchsticks === null || matchsticks.length === 0)
        return false;

    let perimeter = matchsticks.reduce((partialSum, a) => partialSum + a, 0)
    possibleSumSide = perimeter / 4;

    if (perimeter % 4 !== 0)
        return false

    matchsticks.sort((a, b) => a - b)
    matchsticks = matchsticks.reverse()

    sums = new Array(4);
    sums.fill(0);

    return dfs(matchsticks, 0);
};
var dfs = (matchsticks, idx) => {
    // Base
    if (idx === matchsticks.length) {
        return sums[0] === sums[1] && sums[1] === sums[2] && sums[2] === sums[3];
    }

    // Logic
    let ele = matchsticks[idx];
    for (let i = 0; i < 4; i++) {
        if (sums[i] + ele <= possibleSumSide) {
            sums[i] += ele;
            if (dfs(matchsticks, idx + 1)) {
                return true;
            }
            sums[i] -= ele;
        }
    }
    return false;
}

File: problem2.cpp
/*
// Time Complexity : Exponential
// Space Complexity : O(N) size of the stack
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
*/

#include<iostream>
#include<vector>

using namespace std;

class Solution {
    int max_size;
    int arr_size;
    vector<int> res{};
    bool flag{false};
    void display(const vector<int>& v1){
        for(const int& v:v1){
            cout<<v<<"\t";
        }
        cout<<endl;
    }
    void dfs(int idx,vector<int>& matchsticks){
        if(flag) return;
        if(idx == matchsticks.size()){
            //display(res);
            for(int i{};i<arr_size-1;++i){
                if(res.at(i)!=res.at(i+1)){
                    return;
                }
            }
            flag = true;
            return;
        }
        for(int i{};i<arr_size;++i){
            //action
            if(res.at(i)+matchsticks.at(idx)<=max_size){
                res.at(i) = res.at(i) + matchsticks.at(idx);
                //recurse
                dfs(idx+1,matchsticks);
                //backtrack
                res.at(i)= res.at(i) - matchsticks.at(idx);
            }
            
        }
    }
public:
    bool makesquare(vector<int>& matchsticks) {
        sort(matchsticks.begin(),matchsticks.end());
        reverse(matchsticks.begin(),matchsticks.end());
        arr_size = 4;
        res.resize(arr_size,0);
        int sum{};
        for(int i{};i<matchsticks.size();++i){
            sum+=matchsticks.at(i);
        }
        if(sum%4!=0) return false;
        max_size = sum/4;
        int idx{};
        dfs(0,matchsticks);
        return flag;
    }
};